Composite products support for IsProductSaleable.

novikor opened the pull request:
<!---
    Thank you for contributing to Magento Inventory.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
Composite products support for IsProductSalable.

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/inventory#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->


### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/inventory#3200: [WIP] Composite products support for IsProductSaleable.

novikor committed:
Composite products support for IsProductSaleable.

novikor committed:
Composite products support for IsProductSaleable.

novikor commented:
@magento run all tests

swnsma commented:
@magento run all tests

novikor commented:
@magento run Static Tests

novikor commented:
@magento run all tests

novikor commented:
@magento run all tests

Stepa4man commented on InventoryConfigurableProduct/Model/IsProductSalableCondition/IsConfigurableProductSalable.php:
```diff
@@ -0,0 +1,84 @@
+                $results = $this->areProductsSalable->execute($skus, $stockId);
+                foreach ($results as $result) {
+                    if ($result->isSalable()) {
+                        $status = true;
+                        break;
```
`return true`

Stepa4man commented on InventoryConfigurableProduct/Model/IsProductSalableCondition/IsConfigurableProductSalable.php:
```diff
@@ -0,0 +1,84 @@
+        AreProductsSalableInterface $areProductsSalable,
+        Configurable $configurable
+    ) {
+        $this->repository = $repository;
```
Rename to something clear

Stepa4man commented on InventoryConfigurableProduct/Model/IsProductSalableCondition/IsConfigurableProductSalable.php:
```diff
@@ -0,0 +1,84 @@
+    ) {
+        $this->repository = $repository;
+        $this->areProductsSalable = $areProductsSalable;
+        $this->configurable = $configurable;
```
Same here

Stepa4man commented on InventoryConfigurableProduct/Model/IsProductSalableCondition/IsConfigurableProductSalable.php:
```diff
@@ -0,0 +1,84 @@
+        try {
+            $status = false;
+            $product = $this->repository->get($sku);
+            if ($product->getTypeId() === Configurable::TYPE_CODE) {
```
Handle negative case first.

Stepa4man commented on InventoryConfigurableProduct/Model/IsProductSalableCondition/IsConfigurableProductSalable.php:
```diff
@@ -0,0 +1,84 @@
+    {
+        try {
+            $status = false;
+            $product = $this->repository->get($sku);
```
Make DB call after possible negative cases handled and you are sure no chance that you don't use the result of it.

Stepa4man commented on InventoryConfigurableProduct/Model/IsProductSalableCondition/IsConfigurableProductSalable.php:
```diff
@@ -0,0 +1,84 @@
+                    }
+                }
+            }
+        } catch (NoSuchEntityException $e) {
```
I don't like that if there is no product with such sku we return false instead of no such entity exception. What is the reason for this?

Stepa4man commented on InventoryConfigurableProduct/Model/IsProductSalableCondition/IsConfigurableProductSalable.php:
```diff
@@ -0,0 +1,84 @@
+use Magento\Framework\Exception\NoSuchEntityException;
+use Magento\InventorySalesApi\Api\AreProductsSalableInterface;
+use Magento\InventorySalesApi\Api\IsProductSalableInterface;
+
+class IsConfigurableProductSalable implements IsProductSalableInterface
```
Doc block is missed I gues...

novikor replied:
It is not required if the class name describes its purpose well.

Stepa4man commented on InventoryBundleProduct/Model/IsProductSalableCondition/IsBundleSaleableCondition.php:
```diff
@@ -0,0 +1,77 @@
+                    $stockId
+                );
+            }
+        } catch (LocalizedException $e) {
```
So general exception was caught... 
What is the reason for this?

Stepa4man commented on InventoryBundleProduct/Model/IsProductSalableCondition/IsBundleSaleableCondition.php:
```diff
@@ -0,0 +1,77 @@
+        $status = false;
+        try {
+            $product = $this->repository->get($sku);
+            if ($product->getTypeId() === Type::TYPE_CODE) {
```
The negative case should be handled first

Stepa4man commented on InventoryBundleProduct/Model/IsProductSalableCondition/IsBundleSaleableCondition.php:
```diff
@@ -0,0 +1,77 @@
+use Magento\Framework\Exception\LocalizedException;
+use Magento\InventoryBundleProduct\Model\GetBundleProductStockStatus;
+use Magento\InventorySalesApi\Api\IsProductSalableInterface;
+
+class IsBundleSaleableCondition implements IsProductSalableInterface
```
Dock block?

novikor replied:
@Stepa4man, it is not required if the class name describes its purpose well.

Stepa4man commented on InventoryBundleProduct/Model/IsProductSalableCondition/IsBundleSaleableCondition.php:
```diff
@@ -0,0 +1,77 @@
+        GetBundleProductStockStatus $getBundleProductStockStatus
+    ) {
+        $this->bundleProductType = $type;
+        $this->repository = $repository;
```
Clear name, please

Stepa4man reviewed (changes requested):
Hi @novikor 
Thanks for your contribution.

Please check out my comments.

Commit:
Refactoring.
Apply code review changes.

Commit:
Fix static tests.

Commit:
Move product load after type discovery.

Commit:
Fix typo.

Commit:
Refactoring.
Move bundle status check to plugin.
Move configurable plugin to around.
Both facts comes from fact for bundle and configurable produts, salability status is depends only on salability status of their children (and does not depend on the condition chain).

Commit:
Fix static tests, fix integration tests, code review fixes, revert unneded changes.

novikor committed:
Composite products support for IsProductSaleable.
Fixed case when composite product has out of stock status.

novikor committed:
Merge remote-tracking branch 'origin/composite-products-support-for-isProductSalable' into composite-products-support-for-isProductSalable-v2

Commit:
magento/inventory#3140:
Fix static tests.

novikor commented:
@magento create issue

swnsma committed:
Merge branch '1.2-develop' into composite-products-support-for-isProductSalable

novikor committed:
Merge pull request #3151 from magento/composite-products-support-for-isProductSalable-v2

Refactoring.
